Add save functionality to Storage form + build ActionButton

Caleb-Ellis opened the pull request:
## Done
- Added save functionality to Storage form
- Reworked Storage form to use FormikField component
- Updated Storage form styling (no title, left-aligned button)
- Added ActionButton component which wraps Button. It takes props to determine whether it's in a loading or success state. It takes an explicit width prop which is a bit crap but I couldn't figure out how to store button width in the hooks world and figured this works fine for now :man_shrugging: 
- Added some hooks that essentially "fake" a delay after clicking Save, even though saving happens almost instantly and therefore saving state is `false`. Success state is simply determined as "saving state moved from true => false" so will probs need some handling for when that happens due to an error.

## QA
- Go to Storage and check that the button is disabled
- Change some values and check that button is no longer disabled
- Change values back to original (without saving) and check that button becomes disabled again
- Save some changes and check that the button text is replaced by a spinner, and then after a short time it changes to a tick
- Refresh the page and check that state has persisted

## Screenshots
### No values changed
![Screen Shot 2019-08-12 at 18 51 44-fullpage](https://user-images.githubusercontent.com/25733845/62882580-8510dc80-bd32-11e9-8e1e-a691d99c6d04.png)


### Values changed
![Screen Shot 2019-08-12 at 18 51 54-fullpage](https://user-images.githubusercontent.com/25733845/62882589-880bcd00-bd32-11e9-8622-f0153f8657bd.png)


### Saving
![Screen Shot 2019-08-12 at 18 52 00-fullpage](https://user-images.githubusercontent.com/25733845/62882596-8b06bd80-bd32-11e9-8ea4-ba81fc733220.png)


### Saved
![Screen Shot 2019-08-12 at 18 52 01-fullpage](https://user-images.githubusercontent.com/25733845/62882600-8e01ae00-bd32-11e9-97d5-e3176af7c88c.png)

webteam-app commented:
Starting demo at: https://maas-ui-canonical-web-and-design-pr-63.run.demo.haus/

squidsoup commented on ui/src/app/settings/actions/config.js:
```diff
@@ -13,4 +13,18 @@ config.fetch = () => {
+      actionType: "SAVE_CONFIG",
+      message: {
+        method: "config.update",
+        params: { name, value },
```
To answer your first question TODO question, fairly certain this can be an array.

Caleb-Ellis replied:
@squidsoup Hmm yeah I wasn't sure about that. This is the update method in the config websocket handler, which seems to expect a dict with name and value keys:

``` python
    def update(self, params):
        """Update a config value."""
        if not self.user.is_superuser:
            raise HandlerPermissionError()
        if 'name' not in params:
            raise HandlerPKError("Missing name in params")
        if 'value' not in params:
            raise HandlerValidationError("Missing value in params")
        name = params['name']
        value = params['value']
        form = get_config_form(name, {name: value})
        if form.is_valid():
            try:
                request = HttpRequest()
                request.user = self.user
                form.save(ENDPOINT.UI, request)
            except ValidationError as e:
                self._fix_validation_error(name, e.error_dict)
                raise HandlerValidationError(e.error_dict)
            return self._include_choice({
                'name': name,
                'value': Config.objects.get_config(name),
            })
        else:
            self._fix_validation_error(name, form.errors)
            raise HandlerValidationError(form.errors)
```

Unless I'm missing something (highly possible)?

squidsoup replied:
Yep, the API is a bit broken here, but we should not be dispatching multiple actions, the payload should still be an array - we'll just have to handle that in the saga.

squidsoup replied:
Have filed [a bug](https://bugs.launchpad.net/maas/+bug/1839788).

squidsoup replied:
@Caleb-Ellis support for [params as an array landed](https://github.com/canonical-web-and-design/maas-ui/commit/d71059c4b9709ae0b50c91c4cb7428322f43fb0e).

huwshimi commented on ui/src/app/settings/containers/Storage/Storage.js:
```diff
@@ -42,76 +69,70 @@ const Storage = () => {
             <Form onSubmit={handleSubmit}>
               <div className="row">
                 <div className="col-8">
                   <Select
```
These can use the new Formik field component, see: https://github.com/canonical-web-and-design/maas-ui/pull/62#discussion_r312314179

lilyvidenova commented:
@Caleb-Ellis Could you try increasing the time of the active state by 100ms? I feel like it's a little bit too quick.

steverydz commented on ui/src/app/base/hooks.js:
```diff
@@ -35,3 +38,40 @@ export const useLocation = () => {
+  const prevSaving = usePrevious(saving);
+
+  // Temporarily set success to true if saving changes from true to false
+  // TODO: Add a check for errors too
```
Is there a ticket for this? Might be good to put a link to it here

steverydz commented on ui/src/app/base/components/ActionButton/ActionButton.js:
```diff
@@ -0,0 +1,60 @@
+  className,
+  loading,
+  success,
+  width,
```
Why would we need to define a width? Shouldn't this be defined by the text?

Caleb-Ellis commented:
@lilyvidenova Fixed. Changed form to col-6, moved icon to help text, removed yellow border

![Screen Shot 2019-08-13 at 11 48 03-fullpage](https://user-images.githubusercontent.com/25733845/62932574-af10e000-bdc0-11e9-9124-1db59cf14bc4.png)
